L68-core-callback-api.md

```diff
@@ -0,0 +1,106 @@
+Core callback-based completion queue
+----
+* Author(s): vjpai
+* Approver: markdroth
+* Status: Proposed
+* Implemented in: https://github.com/grpc/grpc/projects/12
+* Last updated: June 1, 2020
+* Discussion at
+
+## Abstract
+
+Provide a completion-queue implementation for gRPC core in which the
+completion of an operation triggers the invocation of a registered
+piece of user code rather than posting a tag on an actual queue. The
+new completion queue will have no explicit retrieval function at the
+surface API level.
+
+## Background
+
+gRPC Core uses a completion queue as a notification mechanism of
+completion of RPC stream operations.  `grpc_call_start_batch` (the
+surface API for initiating RPC operation batches) and certain other
+functions associate an RPC operation
+with a _tag_ (currently-unique ID) when it starts. The tag is of type
+`void*` but is not interpreted by gRPC core. The core API user
+(a gRPC wrapped language implementation) will later see that tag when
+accessing the completion queue via a specified accessor function.
+There have long been 2 completion
+queue types:
+
+* `GRPC_CQ_NEXT` : arbitrary number of operations may be queued to it,
+accessed through `grpc_completion_queue_next` which returns the "next"
+item in the queue (but not necessarily FIFO with respect to wall-clock
+time, initiation order, etc.)
+* `GRPC_CQ_PLUCK` : limited number of operations may be queued to it,
+accessed through `grpc_completion_queue_pluck` which allows waiting
+for a specific tag to arrive at the queue and provides that specific
+tag
+
+The completion queue accessor functions may block until a specified
+deadline, and during that time the user's thread is available for the
+library to use in performing RPC operations or polling the operating
+system for the completion of operations.
+
+### Related proposals
+
+* The [EventManager](https://www.github.com/grpc/proposal/XXX) is
+critical to the operation of the callback completion queue since this
+completion queue type will not provide any accessor function to drive
+the progress of the library
+* The callback completion queue is the implementation mechanism for the [C++ callback API](https://www.github.com/grpc/proposal/pull/180)
+
+## Proposal
+
+gRPC core will now support a new completion queue type called
+`GRPC_CQ_CALLBACK`. Although this is called a completion queue because
+it supports the semantic interface of `grpc_completion_queue`,
+operations in this will never be queued in this structure. Instead, a
+completion will cause a user-specified function to be called.
+
+The tags injected into this completion queue
+are still of type `void*` so as not to affect the API of RPC operation
+initiation functions, but these will be interpreted by the
+library. These tags *must* point to a struct of type
+`grpc_completion_queue_functor` which has two user-provided members:
```

Comment:
Is this type exposed in the C-core API?  Are wrapped languages expected to use this?

More generally, if the callback completion queue is required to use EventManager, then what work is needed to get wrapped languages using EventManager?  @muxi @guantaol

Reply:
Is it the other way around that callback completion queue depends on EventManager? If an application enables event manager, it can use callback completion queue or other completion queues. But if the application disables event manager, it can only use next/pluck completion queues.

Python used the callback completion queue directly. The exposed API in this doc is sufficient, with one of the asynchronous mechanism to manage I/O (asyncio, event manager, etc.).

Reply:
TIL that Python uses the callback CQ. Cool.

Reply:
@markdroth - `grpc_completion_queue_functor` is core surface API and will be used by wrapped languages as well. It is a C89 struct.